### PR TITLE
Move logic from `initialize_object` to object store class

### PR DESCRIPTION
**Pull request by hanlint**
This PR moves some of the environment loading logic from the hparams class into the object store proper, while retaining backwards compatibility.

With this change, some additional hparams-specific tests can also be moved into the `test/hparams` folder.

This PR also adds a `dist.barrier()` to `Trainer.close()` which will wait for all ranks to close before exiting.

### CODE REVIEW

File: composer/utils/object_store/object_store_hparams.py [hunk 1]
```diff
@@ -161,20 +161,24 @@ def get_object_store_cls(self) -> Type[ObjectStore]:
...
+            provider_kwargs['port'] = self.port
+        provider_kwargs.update(self.extra_init_kwargs)
+
+        return {
             'provider': self.provider,
             'container': self.container,
-            'provider_kwargs': {},
+            'chunk_size': self.chunk_size,
+            'key_environ': self.key_environ,
+            'secret_environ': self.secret_environ,
```
Comment (mvpatel2000) [inline composer/utils/object_store/object_store_hparams.py:179]: why are these moved out of provider

**Review by mvpatel2000 — APPROVED** [review]
LGTM but not super familiar with this

File: composer/utils/object_store/object_store_hparams.py (hunk 1 above)
Comment (hanlint) [reply composer/utils/object_store/object_store_hparams.py:179]: They weren't -- originally those were used inside `initialize_object` to read from envvars if they exist and populate `provider`. That functionality is now in the object store's own `__init__`.

Comment (mvpatel2000) [reply composer/utils/object_store/object_store_hparams.py:179]: Right, so before they were added to the `provider` object but now they are moved outside `provider` into the parent json (if im understanding this correctly). Just making sure this is a conscious choice

Comment (hanlint) [reply composer/utils/object_store/object_store_hparams.py:179]: Ah, let me clarify!

The `provider` object has not been changed. Before/after it is still `{'key': ..., 'secret': ..}`.

Pseudo-code before:
```
provider = {
'key', ...,
'secret',... }

if key_environ in os.environ:
    provider['key'] = # read from envvar

# create object
return ObjectStore( provider..)

```

After:
```
provider = {
'key', ...,
'secret',... }


# create object
return ObjectStore( provider, key_envirion)

class ObjectStore

def __init__(provider, key_envirion):

    # logic now lives here
    if key_environ in os.environ:
        provider['key'] = ...

```

**Comment by mvpatel2000** [comment]
Discussed offline, things are clear to me. Confident in lgtm
